### PR TITLE
add callbacks to process notification delivery/err

### PR DESCRIPTION
**Pull request by sdowell**
This is intended to provide a generic framework for processing notification delivery/error events.

For example, a user could instantiate a controller with some business logic to monitor when notifications are misconfigured or fail to send to a particular service.

Issues: https://github.com/argoproj/notifications-engine/issues/135

### CODE REVIEW

File: pkg/controller/controller.go [hunk 1]
```diff
@@ -160,16 +207,27 @@ func (c *notificationController) processResource(resource v1.Object, logEntry *l
...
+						AlreadyNotified: true,
+					})
 				} else {
 					logEntry.Infof("Sending notification about condition '%s.%s' to '%v'", trigger, cr.Key, to)
 					if err := api.Send(un.Object, cr.Templates, to); err != nil {
 						logEntry.Errorf("Failed to notify recipient %s defined in resource %s/%s: %v",
 							to, resource.GetNamespace(), resource.GetName(), err)
 						notificationsState.SetAlreadyNotified(trigger, cr, to, false)
 						c.metricsRegistry.IncDeliveriesCounter(trigger, to.Service, false)
+						eventSequence.addError(fmt.Errorf("failed to deliver notification %s to %s: %v", trigger, to.Service, err))
```
Comment (nan-yu) [inline pkg/controller/controller.go:222]: Suggest including the entire `to` object in the log, which includes both `to.Service` and `to.Recipient`.

Comment (sdowell) [reply pkg/controller/controller.go]: Done

**Review by nan-yu — APPROVED** [review]
/lgtm

**Comment by pasha-codefresh** [comment]
Hey @nan-yu @sdowell i will check it tomorrow and hopefully merge . Sorry for delay

**Comment by nan-yu** [comment]
Thanks a lot! @pasha-codefresh

**Comment by pasha-codefresh** [comment]
Working on review

**Comment by nan-yu** [comment]
Thanks again @pasha-codefresh !

We have two more questions regarding the project: 
- What is the release cadence of the notifications engine library?
- What is the best way to find the reviewers. We have been tagging you and @alexmt . It would be good to have an owners and reviewers file.

**Comment by pasha-codefresh** [comment]
If you want use this code in argo-cd , you just need open PR in argo-cd and also you can assign it to me 

Yeah, we need add reviewers file, i will take care of it . Yes @alexmt and me doing reviews here

**Pull request merged**